### PR TITLE
PMM-3595 Changes for qan-mysql-perfschema.

### PR DESCRIPTION
**Pull request by AlekSi**

### CODE REVIEW

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

File: api/qan/collector.proto [hunk 1]
```diff
@@ -2,56 +2,89 @@ syntax = "proto3";
...
-  string agent_uuid = 9; // Identifier of agent that collect and send metrics.
-  int64 period_start = 10; // Time when collection of bucket started.
-  uint32 period_length = 11; // Duration of bucket.
-  string metrics_source = 12; // Ex.: slowlog, perf schema, Mongo DB, Percona Server, etc.
-  string example = 13; // One of query example from set found in bucket.
+  // PostgreSQL database.
+  string d_database = 4;
+  // MySQL database; PostgreSQL schema; MongoDB collection.
+  // FIXME(askomorokhov) why collection?!
+  string d_schema = 5;
```
Comment (AlekSi) [inline api/qan/collector.proto]: @askomorokhov ^

File: api/qan/collector.proto [hunk 2]
```diff
@@ -2,56 +2,89 @@ syntax = "proto3";
...
+  // Custom labels names:values.
+  map<string, string> labels = 8;
+  // Identifier of agent that collect and send metrics.
+  string agent_uuid = 9;
+  // Time when collection of bucket started.
+  // FIXME(askomorokhov) what unit? why not google.protobuf.Timestamp?
+  int64 period_start = 10;
+  // Duration of bucket.
+  // FIXME(askomorokhov) what unit? why not google.protobuf.Duration?
+  uint32 period_length = 11;
```
Comment (AlekSi) [inline api/qan/collector.proto]: @askomorokhov ^

File: api/qan/collector.proto [hunk 3]
```diff
@@ -2,56 +2,89 @@ syntax = "proto3";
...
+  // Ex.: slowlog, perf schema, MongoDB, Percona Server, etc.
+  // FIXME(askomorokhov) Percona Server? why not enum?
+  string metrics_source = 12;
```
Comment (AlekSi) [inline api/qan/collector.proto]: @askomorokhov ^

File: api/qan/collector.proto [hunk 4]
```diff
@@ -2,56 +2,89 @@ syntax = "proto3";
...
   ExampleFormat example_format = 14;
-  uint32 is_truncated = 15; // Indicates if query examples is too long and was truncated.
+  // Indicates if query examples is too long and was truncated.
+  // FIXME(askomorokhov) why not boolean?
+  uint32 is_truncated = 15;
   ExampleType example_type = 16;
-  string example_metrics = 17; // JSON raw metrics for given query example.
+  // JSON raw metrics for given query example.
+  // FIXME(askomorokhov) wat?
+  string example_metrics = 17;
```
Comment (AlekSi) [inline api/qan/collector.proto]: @askomorokhov ^

File: api/qan/collector.proto [hunk 5]
```diff
@@ -2,56 +2,89 @@ syntax = "proto3";
...
+  // Ex.: slowlog, perf schema, MongoDB, Percona Server, etc.
+  // FIXME(askomorokhov) Percona Server? why not enum?
+  string metrics_source = 12;
+  // One of query example from set found in bucket.
+  string example = 13;
   ExampleFormat example_format = 14;
-  uint32 is_truncated = 15; // Indicates if query examples is too long and was truncated.
+  // Indicates if query examples is too long and was truncated.
+  // FIXME(askomorokhov) why not boolean?
+  uint32 is_truncated = 15;
```
Comment (AlekSi) [inline api/qan/collector.proto]: @askomorokhov ^

File: api/qan/collector.proto [hunk 6]
```diff
@@ -137,9 +170,15 @@ message MetricsBucket {
   float m_tmp_table_sizes_min = 110;
   float m_tmp_table_sizes_max = 111;
   float m_tmp_table_sizes_p99 = 112;
+  //
   // boolean metrics
+  //
+  // FIXME(askomorokhov) why they are float then?
```
Comment (AlekSi) [inline api/qan/collector.proto]: @askomorokhov ^

File: api/qan/collector.proto [hunk 7]
```diff
@@ -2,56 +2,89 @@ syntax = "proto3";
...
+  string d_username = 6;
+  // client IP or hostname.
+  string d_client_host = 7;
+  // Custom labels names:values.
+  map<string, string> labels = 8;
+  // Identifier of agent that collect and send metrics.
+  string agent_uuid = 9;
+  // Time when collection of bucket started.
+  // FIXME(askomorokhov) what unit? why not google.protobuf.Timestamp?
+  int64 period_start = 10;
```
Comment (AlekSi) [inline api/qan/collector.proto]: @askomorokhov ^

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

File: api/qan/collector.proto (hunk 7 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: This is unix timestamp in seconds. Considered this arguments: https://stackoverflow.com/questions/3574716/date-and-time-type-for-use-with-protobuf#3627777
I see more disadvantages of using `google.protobuf.Timestamp` here.

File: api/qan/collector.proto (hunk 2 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: seconds. For now it should be 60 sec. In future it is can change to another duration.

File: api/qan/collector.proto (hunk 3 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: I thing it should be changed to enum.

File: api/qan/collector.proto (hunk 5 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: Agree, let's change to boolean.

File: api/qan/collector.proto (hunk 4 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: Please correct comment to `Metrics of given query example in JSON format.`

File: api/qan/collector.proto (hunk 6 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: This is boolean metrics in MySQL Percona Server slow log.
Values passes in gRCP will be int.
We desided to store all metrics in ClickHouse as float32.
What is your opinion? Does we win more in speed or memory with additional converting  int<->float.?

File: api/qan/collector.proto (hunk 4 above)
Comment (AlekSi) [reply api/qan/collector.proto]: I still don't get what it means.
Besides, we are sending a well structured message with strongly typed fields – and with any JSON object inside?

File: api/qan/collector.proto (hunk 6 above)
Comment (AlekSi) [reply api/qan/collector.proto]: `float32(1)` is like 100x faster than float decoding in protobuf.

Comment (AlekSi) [reply api/qan/collector.proto]: But that metrics are actually boolean? Count? Sum? That's unlikely :)

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by AlekSi** [comment]
@DexterHD Please check agent/inventory part.

File: api/qan/collector.proto (hunk 4 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: This JSON object will simply displayed as table on UI below query examples without any aggregations.

Comment (AlekSi) [reply api/qan/collector.proto]: And what it will contain?

File: api/agent/agent.go [hunk 8]
```diff
@@ -1,4 +1,4 @@
-package agent
+package agentpb
```
Comment (DexterHD) [inline api/agent/agent.go]: @AlekSi What does it mean `pb`?

Comment (AlekSi) [reply api/agent/agent.go]: ProtoBuf. We are slowly moving to the best practices we ignored before: https://github.com/percona/pmm/blob/696407ba1a1c11140a3da6dbf408a53fa0b314eb/api/prototool.yaml#L27 Without it we often have a clash of package names.

Comment (AlekSi) [reply api/agent/agent.go]: >And why directory and filename is an agent but package name is agentpb?

That's the next step.

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

File: api/qan/collector.proto (hunk 6 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: > float32(1) is like 100x faster than float decoding in protobuf. 

sorry, did not get this your comment.
For me at a glance they seems quite similar in speed:
https://developers.google.com/protocol-buffers/docs/encoding
https://github.com/golang/protobuf/blob/master/proto/decode.go

> But that metrics are actually boolean? Count? Sum? That's unlikely :)

Let's  imaging that in slow log we have 10 entry of same query.
So, num_query = 10. if query use Query cache that QC_Hit will be present in slowlog entry.
```
# QC_Hit: No  Full_scan: Yes  Full_join: No  Tmp_table: No  Tmp_table_on_disk: No
# Filesort: No  Filesort_on_disk: No  Merge_passes: 0
```
or 
```
# QC_Hit: Yes  Full_scan: Yes  Full_join: No  Tmp_table: No  Tmp_table_on_disk: No
# Filesort: No  Filesort_on_disk: No  Merge_passes: 0
```
Hence m_qc_hit_cnt is counter how many time QC_Hit "flag" appear in given segment of slowlog. Zero is NULL here.
And m_qc_hit_sum is counter how many time QC_Hit was equal to "Yes".

File: api/qan/collector.proto (hunk 4 above)
Comment (askomorokhov) [reply api/qan/collector.proto]: all data from slow log entry for specific query.
Ex.:
```
# User@Host: mailboxer[mailboxer] @  [192.168.10.165]
# Thread_id: 11167745  Schema: board
# Query_time: 1.009400  Lock_time: 0.000190  Rows_sent: 4  Rows_examined: 1543719  Rows_affected: 0  Rows_read: 4
# Bytes_sent: 278  Tmp_tables: 0  Tmp_disk_tables: 0  Tmp_table_sizes: 0
# InnoDB_trx_id: 1500
# QC_Hit: No  Full_scan: Yes  Full_join: No  Tmp_table: No  Tmp_table_on_disk: No
# Filesort: No  Filesort_on_disk: No  Merge_passes: 0
#   InnoDB_IO_r_ops: 6415  InnoDB_IO_r_bytes: 105103360  InnoDB_IO_r_wait: 0.001279
#   InnoDB_rec_lock_wait: 0.000000  InnoDB_queue_wait: 0.000000
#   InnoDB_pages_distinct: 6430
SET timestamp=1346844943;
SELECT id,title,production_year FROM title WHERE title = 'Bambi';
```

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Comment by JNKPercona** [comment]
http://docs-test.cd.percona.com/PR-289/

**Pull request closed without merging**